# Coding Exercises (Part 1)

Now, you will have the opportunity to practice what you have learned. <br>
__Follow the instructions__ and insert your code! 

The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need some help or if you want to check your code, you can also have a look at the __solutions__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Exercise 3: Integers, Floats, Lists and built-in functions

1. __Create and save objects__ of your choice with the following __data types__. Check the data type with the appropriate function!

integer

In [2]:
integer = 1
print(integer)
type(integer)

1


int

float

In [3]:
float = 1.05
print(float)
type(float)

1.05


float

string

In [4]:
string = "this is a string"
print(string)
type(string)

this is a string


str

bool

In [5]:
bool = True
print(bool)
type(bool)

True


bool

2. Determine the __sum__, the __number of elements__ and __maximum value__ in list l. Use __built-in functions__!

In [6]:
# run the cell
l = [-300, 40, 140, -430, 85, 280, 0, -560, 340]

In [7]:
#sum
sum(l)

-405

(result: -405)

In [10]:
#number of elements
len(l)

9

(result: 9)

In [12]:
#maximum
max(l)

340

(result: 340)

3. Determine the __number of elements__ and __the alphabetically first value__ in string s. Use __built-in functions__!

In [13]:
#run the cell
s = "myrmecophilous"

In [14]:
#number of elements
len(s)

14

(result: 14)

In [16]:
#alphabetically first value
sorted_s = sorted(s)
print(sorted_s)

['c', 'e', 'h', 'i', 'l', 'm', 'm', 'o', 'o', 'p', 'r', 's', 'u', 'y']


(result: "c")

4. __Check__ whether the following equation is __True__ in Python: 0.2 + 0.2 + 0.2 = 0.6

In [21]:
a = 0.2 + 0.2 + 0.2 
b = 0.6

if a == b:
    print("True")
else:
    print("False")

False


5. Get the __first 20 decimals__ of 0.2 and 0.6 and explain the result of exercise 4.

In [25]:
format(0.2, ".20f")

'0.20000000000000001110'

(result: '0.20000000000000001110')

In [26]:
format(0.6, ".20f")

'0.59999999999999997780'

(result: '0.59999999999999997780')

6. __Round__ 1.56456 to __2 decimals__! Check the __first 20 decimals__ of the rounded float. Does rounding guarantee exact numbers in Python?

In [32]:
c = round(1.56456, 2)
print(c)
format(c, ".20f")

1.56


'1.56000000000000005329'

(result: 1.56)

(result: '1.56000000000000005329')

7. __Round__ 2,320,576 to the next __multiple of 1,000__!

(result: 2321000)

We have the list cf with cashflows starting today until year 6:

In [33]:
# run the cell
cf = [-2000, 200, 300, 500, 750, 1000, 1500]
cf

[-2000, 200, 300, 500, 750, 1000, 1500]

8. Create the following __slices__ of cf:

first two elements

In [35]:
cf[0:2]

[-2000, 200]

the elements 300, 500 and 750

In [ ]:
cf[2]

the last three elements

In [38]:
cf[-3:]

[750, 1000, 1500]

all elements, except the last element

In [39]:
cf[:6]

[-2000, 200, 300, 500, 750, 1000]

every second element (starting from first element)

In [40]:
cf[0:len(cf):2]

[-2000, 300, 750, 1500]

9. __Add__ the cashflow 1200 for year 7 to list cf. __Change__ cf in memory!

In [44]:
cf_new = cf + [1200]
print(cf_new)

[-2000, 200, 300, 500, 750, 1000, 1500, 1200]


10. __Change__ the cashflow in t2 (300) to 400. __Change__ cf in memory!

In [46]:
cf_new[2] = 400
print(cf_new)

[-2000, 200, 400, 500, 750, 1000, 1500, 1200]


11. __Remove__ the last three years of cf (with slicing and variable reassignment )

In [48]:
cf_remove = cf_new[:-3]
print(cf_remove)

[-2000, 200, 400, 500, 750]


12. __Add 20__ to each and every element of cf. Create the new list __cf_new__!

In [55]:
cf_new = []
for i in cf:
    cf_new.append(i + 20)
print(cf_new)

[-1980, 220, 320, 520, 770, 1020, 1520]


(result: [-1980, 220, 420, 520, 770])

13. __Sort__ cf_new from __high to low__! __Change__ cf_new in memory!

In [56]:
cf_sort = cf_new

In [57]:
cf_sort.sort()

(result: [770, 520, 420, 220, -1980])

Initially, we have two projects with identical cashflows:

In [60]:
#run the cell
cf1 = [-50, 10, 30, 50]
cf2 = cf1

Then, the first cashflow of project 1 changes to -60.

In [61]:
#run the cell
cf1[0] = -60
cf1

[-60, 10, 30, 50]

14. Check, whether __cf2 changed__ as well!

In [63]:
print(cf2)

[-60, 10, 30, 50]


15. __Rewrite__ the above code and make sure that __cf2 does not change__! Follow __best practices__!

In [70]:
cf1 = [-50, 10, 30, 50]
cf2 = cf1
print(cf1, cf2)

[-50, 10, 30, 50] [-50, 10, 30, 50]


In [71]:
cf1[0] = -60
print(cf1, cf2)

[-60, 10, 30, 50] [-60, 10, 30, 50]


In [72]:
cf1 = cf1 + [15]
print(cf1, cf2)

[-60, 10, 30, 50, 15] [-60, 10, 30, 50]


# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Solutions (Stop here if you want to code on your own!)

# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

1. __Create and save objects__ of your choice with the following __data types__. Check the data type with the appropriate function!

integer

In [ ]:
a = 14
a

In [ ]:
type(a)

float

In [ ]:
a = 12.56
a

In [ ]:
type(a)

string

In [ ]:
a = "Twitter"
a

In [ ]:
type(a)

bool

In [ ]:
a = True
a

In [ ]:
type(a)

2. Determine the __sum__, the __number of elements__ and __maximum value__ in list l. Use __built-in functions__!

In [ ]:
# run the cell
l = [-300, 40, 140, -430, 85, 280, 0, -560, 340]

In [ ]:
#sum
sum(l)

(result: -405)

In [ ]:
#number of elements
len(l)

(result: 9)

In [ ]:
#maximum
max(l)

(result: 340)

3. Determine the __number of elements__ and __the alphabetically first value__ in string s. Use __built-in functions__!

In [ ]:
#run the cell
s = "myrmecophilous"

In [ ]:
#number of elements
len(s)

(result: 14)

In [ ]:
#alphabetically first value
min(s)

(result: "c")

4. __Check__ whether the following equation is __True__ in Python: 0.2 + 0.2 + 0.2 = 0.6

In [ ]:
0.2 + 0.2 + 0.2 == 0.6

5. Get the __first 20 decimals__ of 0.2 and 0.6 and explain the result of exercise 4.

In [ ]:
format(0.2, ".20f")

(result: '0.20000000000000001110')

In [ ]:
format(0.6, ".20f")

(result: '0.59999999999999997780')

6. __Round__ 1.56456 to __2 decimals__! Check the __first 20 decimals__ of the rounded float. Does rounding guarantee exact numbers in Python?

In [ ]:
a = round(1.56456, 2)
a

(result: 1.56)

In [ ]:
format(a, ".20f")

(result: '1.56000000000000005329' -> No!)

7. __Round__ 2,320,576 to the next __multiple of 1,000__!

In [ ]:
round(2320576, -3)

(result: 2321000)

We have the list cf with cashflows starting today until year 6:

In [ ]:
# run the cell
cf = [-2000, 200, 300, 500, 750, 1000, 1500]
cf

8. Create the following __slices__ of cf:

first two elements

In [ ]:
cf[:2]

the elements 300, 500 and 750

In [ ]:
cf[2:5]

the last three elements

In [ ]:
cf[-3:]

all elements, except the last element

In [ ]:
cf[:-1]

every second element (starting from first element)

In [ ]:
cf[::2]

9. __Add__ the cashflow 1200 for year 7 to list cf. __Change__ cf in memory!

In [ ]:
cf.append(1200)

In [ ]:
cf

10. __Change__ the cashflow in t2 (300) to 400. __Change__ cf in memory!

In [ ]:
cf[2] = 400

In [ ]:
cf

11. __Remove__ the last three years of cf (with slicing and variable reassignment )

In [ ]:
cf = cf[:-3]

In [ ]:
cf

12. __Add 20__ to each and every element of cf. Create the new list __cf_new__!

In [ ]:
cf_new = []
for element in cf:
    cf_new.append(element + 20)
cf_new

(result: [-1980, 220, 420, 520, 770])

13. __Sort__ cf_new from __high to low__! __Change__ cf_new in memory!

In [ ]:
cf_new.sort(reverse = True)

In [ ]:
cf_new

(result: [770, 520, 420, 220, -1980])

Initially, we have two projects with identical cashflows:

In [ ]:
#run the cell
cf1 = [-50, 10, 30, 50]
cf2 = cf1

Then, the first cashflow of project 1 changes to -60.

In [ ]:
#run the cell
cf1[0] = -60
cf1

14. Check, whether __cf2 changed__ as well!

In [ ]:
cf2

it changed!

15. __Rewrite__ the above code and make sure that __cf2 does not change__! Follow __best practices__!

In [ ]:
cf1 = [-50, 10, 30, 50]
cf2 = cf1.copy()

In [ ]:
cf1[0] = -60
cf1

In [ ]:
cf2